src/Symfony/Component/Serializer/Instantiator/Instantiator.php

```diff
@@ -0,0 +1,119 @@
+{
+    public const DEFAULT_CONSTRUCTOR_ARGUMENTS = AbstractObjectNormalizer::DEFAULT_CONSTRUCTOR_ARGUMENTS;
+
+    use ObjectToPopulateTrait;
+    use DenormalizerAwareTrait;
+
+    private $classDiscriminatorResolver;
+    private $propertyListExtractor;
+    private $nameConverter;
+
+    public function __construct(PropertyListExtractorInterface $propertyListExtractor = null, NameConverterInterface $nameConverter = null)
+    {
+        $this->propertyListExtractor = $propertyListExtractor;
+        $this->nameConverter = $nameConverter;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function instantiate(string $class, $data, $format = null, array $context = [])
+    {
+        $allowedAttributes = $this->propertyListExtractor ? $this->propertyListExtractor->getProperties($class, $context) : null;
+        $reflectionClass = new \ReflectionClass($class);
+        $constructor = $reflectionClass->getConstructor();
+
+        if (null === $constructor) {
+            return new $class();
+        }
+
+        $constructorParameters = $constructor->getParameters();
+
+        $params = [];
+        foreach ($constructorParameters as $constructorParameter) {
+            $paramName = $constructorParameter->name;
+            $key = $this->nameConverter ? $this->nameConverter->normalize($paramName, $class, $format, $context) : $paramName;
+            $allowed = null === $allowedAttributes || \in_array($paramName, $allowedAttributes, true);
+
+            if ($allowed && $constructorParameter->isVariadic() && \array_key_exists($key, $data)) {
+                if (!\is_array($data[$paramName])) {
+                    throw new RuntimeException(sprintf('Cannot create an instance of %s from serialized data because the variadic parameter %s can only accept an array.', $class, $constructorParameter->name));
+                }
+
+                $params = array_merge($params, $data[$paramName]);
+            } elseif ($allowed && \array_key_exists($key, $data)) {
+                $parameterData = $data[$key];
+
+                if (null === $parameterData && $constructorParameter->allowsNull()) {
+                    $params[] = null;
+
+                    continue;
+                }
+
+                $params[] = $this->denormalizeParameter($reflectionClass, $constructorParameter, $paramName, $parameterData, $context, $format);
+            } elseif (\array_key_exists($key, $context[self::DEFAULT_CONSTRUCTOR_ARGUMENTS][$class] ?? [])) {
+                $params[] = $context[self::DEFAULT_CONSTRUCTOR_ARGUMENTS][$class][$key];
+            } elseif ($constructorParameter->isDefaultValueAvailable()) {
+                $params[] = $constructorParameter->getDefaultValue();
+            } else {
+                throw new MissingConstructorArgumentsException(sprintf('Cannot create an instance of %s from serialized data because its constructor requires parameter "%s" to be present.', $class, $constructorParameter->name));
```

Comment:
In case of missing constructor arguments, you probably should fall back on using the property access component to arbitrarily set properties onto the object. Problem is that your code change with time, but serialized entities stored somewhere don't.

Reply:
Just forget the previous comment, sorry for the noise. But it raise an interesting question: the instanciator sets some values using the constructor, then the API using it (for example, the normalizer because it's today the only API using it I know of) will set some others, it seems like each component responsibility is unclear, they both walk on each other foot.

Reply:
Instantiator will only set values related to constructor, if there is other data to set, it will be done after by the ObjectNormalizer

Reply:
How will the normalizer now which properties have already been set or not if the instanciator becomes an opaque interface ? It seems that it will then be forced to re-inject all properties ? Maybe not since you pass an array reference, you can remove constructor-set values from the array, but it's a very weird interface then, you have two different outputs: the object, and the modified array. It's not cognitively-easy. I know do understand your answer below thought.

Reply:
Related to https://github.com/symfony/symfony/pull/30956#discussion_r378252580

Actually we don't have a simpler way to pass both object and new data array, so we use the reference to handle the data. It's probably not the best way to do it but it can be improved. 
I think we should more focus on code duplication in Serializer before focusing on attended behavior even if they do use references.

Reply:
Yes I finally understood why you're doing it this way now. This is weird because the instanciator and the normalizer then share some bits of internal logic then, whereas having an external interface main benefit is "divide and conquer", making every component simpler and avoid cognitive surcharge when reading bare algorithm (also make it simpler to maintain). I agree this is not an easy one, thanks for taking the time to answer.